File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {

    public int[] solution(String msg) {

        List<Integer> answer = new ArrayList<>();
        HashMap<String, Integer> map = new HashMap<>();

        int index = 1;

        // A ~ Z 초기화
        for (; index < 27; index++) {
            map.put(String.valueOf((char) ('A' + index - 1)), index);
        }

        for (int i = 0; i < msg.length(); i++) {

            String temp = String.valueOf(msg.charAt(i));

            int j = i;
            int count = 0;

            while (map.containsKey(temp)) {

                count++;
                j++;

                if (j >= msg.length()) {
                    break;
                }

                temp += msg.charAt(j);
            }

            i += count - 1;

            if (!map.containsKey(temp)) {
                map.put(temp, index++);
                temp = temp.substring(0, temp.length() - 1);
            }

            answer.add(map.get(temp));
        }

        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
}

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.js
function solution(msg) {

    const answer = [];
    const map = new Map();

    let index = 1;

    // A ~ Z 초기화
    for (; index < 27; index++) {
        map.set(String.fromCharCode(64 + index), index);
    }

    for (let i = 0; i < msg.length; i++) {

        let temp = msg[i];

        let j = i;
        let count = 0;

        while (map.has(temp)) {

            count++;
            j++;

            if (j >= msg.length) {
                break;
            }

            temp += msg[j];
        }

        i += count - 1;

        if (!map.has(temp)) {
            map.set(temp, index++);
            temp = temp.slice(0, -1);
        }

        answer.push(map.get(temp));
    }

    return answer;
}